silence/silence.go

```diff
@@ -281,12 +281,12 @@ func (s *Silences) GC() (int, error) {
 	s.mtx.Lock()
 	defer s.mtx.Unlock()
 
-	for id, sil := range s.st {
+	for id, sil := range s.st.data {
 		if sil.ExpiresAt.IsZero() {
 			return n, errors.New("unexpected zero expiration timestamp")
 		}
 		if !sil.ExpiresAt.After(now) {
-			delete(s.st, id)
+			delete(s.st.data, id)
```

Comment:
Reading this now, this is probably what caused the concurrent read/write. If you look at the trace you provided us, it starts with:

```
time="2017-09-08T12:38:05Z" level=info msg="running maintenance" component=silences source="silence.go:232" 
time="2017-09-08T12:38:05Z" level=info msg="running maintenance" component=nflog source="nflog.go:286" 
time="2017-09-08T12:38:05Z" level=info msg="maintenance done" component=nflog duration=137.709µs source="nflog.go:303" 
time="2017-09-08T12:38:05Z" level=info msg="maintenance done" component=silences duration=84.744µs source="silence.go:249" 
fatal error: concurrent map iteration and map write
```

Access to `s.st.data` here also needs to be locked.

I'm a bit curious on the internals of how go works, because access to silences was definitely locked, but it seems the internal map was still concurrently available. Maybe it's something to do with the map being a type alias.

@fabxc @beorn7 any ideas or light you can shed on this?

Reply:
I don't think this was caused by the maintainance, I think it was caused by two `Encode` functions simulatneously encoding the same data:
```
github.com/prometheus/alertmanager/silence.gossipData.Encode(0xc424e69c20, 0xc82580, 0xc424e69c20, 0xc4213f37c0)
github.com/prometheus/alertmanager/silence.gossipData.Encode(0xc424e69c20, 0xc82580, 0xc424e69c20, 0xc4233ab960)
```

Here s.st.data is protected by s.mtx (it's never being accessed without this mutex being held)